Firestore/Example/Tests/Integration/API/FIRDatabaseTests.mm

```diff
@@ -531,6 +531,44 @@ - (void)testAddingToACollectionYieldsTheCorrectDocumentReference {
   [self awaitExpectations];
 }
 
+- (void)testSnapshotsInSyncListenerFiresAfterListenersInSync {
+  FIRCollectionReference *coll = [self.db collectionWithPath:@"collection"];
+  FIRDocumentReference *ref = [coll addDocumentWithData:@{@"foo" : @1}];
+  NSMutableArray<NSString *> *events = [NSMutableArray array];
+
+  XCTestExpectation *gotInitialSnapshot = [self expectationWithDescription:@"gotInitialSnapshot"];
+  __block bool setupComplete = false;
+
+  [ref addSnapshotListener:^(FIRDocumentSnapshot *snapshot, NSError *error) {
```

Comment:
Why do we need this initial snapshot listener?

Reply:
We want to use the snapshot listener to check that the snapshots-in-sync fires after the snapshot listener fires.

Reply:
You're not listening to both concurrently though so this test only verifies that the snapshot listener fires and then when you start a snapshots-in-sync listener that also fires but these are effectively unrelated because we weren't listening to both at the same time.

Reply:
I don't understand @wilhuff's feedback or @thebrianchen's fix.

I think "You're not listening to both concurrently" was incorrect.  We _are_ listening to both concurrently.  We just wait for the first event from the snapshot listener before adding the snapshots-in-sync listener to avoid race conditions between the initial server snapshot (a nonexistent doc snapshot) and the local write.  But they both end up active concurrently.

By removing the gotInitialSnapshot expectation and waiting for it, the test is now race-y since we may or may not get the initial server snapshot before our local write is processed.  If we _don't_ get the server snapshot first, then we'll only end up getting 3 events ["snapshots-in-sync", "doc", "snapshots-in-sync"]

So I think the prior state of this test was correct and the latest change should be undone. Unless I'm missing something.

Reply:
To clarify without arguing any specific point:

  * My initial feedback was that I didn't understand why this was necessary.
  * I was essentially prompting for a comment; a comment like the description you just gave would satisfy my concern.
  * During discussion with Brian he didn't articulate that point of view.
  * So I suggested that if this is what it takes to get a reliable sequence of events we can't expect our users to do this and to the extent we can do so, we should make the tests mirror interactions are users might make

If disabling the synchronization makes the test flaky then this isn't the fix we want.

Secondarily though I don't entirely understand why this test would be flaky. If we have one document listener and it chooses not to raise a snapshot initially, why would we get a snapshots-in-sync event?

Reply:
The race is because:
1. We start a listener which sends a request to the backend for the current data. When that data arrives, we'll raise an event.
2. We perform a local write, which will generate a latency-compensated event (that overrides any server data).

If the backend's response to #1 gets into the AsyncQueue before the local write operation does, then we get 2 events: the backend response, then the local write.  If it doesn't, then we'll only get 1: the local write.

I would expect that as written this test will fail against prod, since the local write will win the race and so we'd only get one event, but (I guess) against the emulator, the backend response wins and we get 2 events, so the test passes.

The "fix" is to add the listener and then immediately wait for the initial snapshot (from the backend) before we proceed with the rest of the test and do our local write.  That way we know we will get one (additional) event.

We have lots of existing tests that do this (add a listener and then immediately wait for the initial empty event), e.g.: https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Integration/API/FIRWriteBatchTests.mm#L208

But adding a comment to make it clearer isn't a bad idea. Something like:

> Wait for the initial event from the backend so that we know we'll get exactly one snapshot event for our local write below.

Reply:
Ah sorry about that Gil. I was struggling to explain the concept behind Michael's comment in my [web pr](https://github.com/firebase/firebase-js-sdk/pull/2100#discussion_r319221356). Surprisingly, I tried running the integration test with synchronization disabled 100 times in a loop several times, and I didn't get any flakes. 

> Secondarily though I don't entirely understand why this test would be flaky. If we have one document listener and it chooses not to raise a snapshot initially, why would we get a snapshots-in-sync event? 

In what situations do we not get a snapshot initially? I thought the snapshot listener will always fire once after it has been added. The snapshots-in-sync listener is scheduled to be fired immediately after it has been added for the first time. On Android, I ran into the following situation (different that what Michael posted):
1. snapshots-in-sync listener fires after it has been added.
2. snapshot listener fires its initial event, causing the snapshots-in-sync listener to fire again.
3. As a result, we get ["snapshots-in-sync", "doc", snapshots-in-sync"] before we write to the doc again.

Given this, is the right approach to leave the synchronization check in place to prevent flakiness?

Reply:
At this point I think you and @mikelehen have worked out how this should work and I'm convinced that the nature of this test could cause flakes (even if we haven't observed them).

My initial comment remains: it's not obvious to an uninitiated reader why the test has to wait for the first document snapshot. Please add a comment that describes the race you're avoiding.